src/realm/query.cpp

```diff
@@ -1822,38 +1822,37 @@ std::string Query::validate()
     return root_node()->validate(); // errors detected by QueryEngine
 }
 
-std::string Query::get_description(util::serializer::SerialisationState& state) const
+const std::string& Query::get_description(util::serializer::SerialisationState& state) const
 {
-    std::string description;
     if (root_node()) {
         if (m_view) {
             throw SerialisationError("Serialisation of a query constrained by a view is not currently supported");
         }
-        description = root_node()->describe_expression(state);
+        m_description_buffer = root_node()->describe_expression(state);
     }
     else {
         // An empty query returns all results and one way to indicate this
         // is to serialise TRUEPREDICATE which is functionally equivalent
-        description = "TRUEPREDICATE";
+        m_description_buffer = "TRUEPREDICATE";
     }
     if (this->m_ordering) {
-        description += " " + m_ordering->get_description(m_table);
+        m_description_buffer += " " + m_ordering->get_description(m_table);
     }
-    return description;
+    return m_description_buffer;
 }
 
-Query& Query::set_ordering(std::unique_ptr<DescriptorOrdering> ordering)
+Query& Query::set_ordering(util::bind_ptr<DescriptorOrdering> ordering)
 {
     m_ordering = std::move(ordering);
     return *this;
 }
 
-std::shared_ptr<DescriptorOrdering> Query::get_ordering()
+util::bind_ptr<DescriptorOrdering> Query::get_ordering()
 {
     return std::move(m_ordering);
```

Comment:
Why are we moving `m_ordering` here, won't that leave m_ordering holding a nullptr? Maybe a better thing to do here is to change the signature of `get_ordering()` to `const util::bind_ptr<DescriptorOrdering>&`?

Reply:
The idea is actually to move the ordering out of the query object.

Reply:
This feels like a confusing API. Methods named "get" that are not only not const, but mutate the thing they're getting that aren't &&-qualified feels like a real anti-pattern at best. Why did we design this this way?